xtp/src/libxtp/dftengine/dftengine.cc

```diff
@@ -510,129 +521,404 @@ bool DFTEngine::EvaluateActiveRegion(Orbitals& orb) {
+      Mat_p_Energy xc_truncated =
+          vxcpotential.IntegrateVXC(TruncatedDensityMatrix);
+      double E_xc_truncated = xc_truncated.energy();
+
+      // get the Hartree contribution for the updated truncated density matrix
+      Eigen::MatrixXd J_truncated = Eigen::MatrixXd::Zero(
+          TruncatedDensityMatrix.rows(), TruncatedDensityMatrix.cols());
+      Eigen::MatrixXd K_truncated = Eigen::MatrixXd::Zero(
+          TruncatedDensityMatrix.rows(), TruncatedDensityMatrix.cols());
+      if (ScaHFX_ > 0) {
+        std::array<Eigen::MatrixXd, 2> JandK_truncated = CalcERIs_EXX(
+            MOs_trunc.eigenvectors(), TruncatedDensityMatrix, 1e-12);
+        J_truncated = JandK_truncated[0];
+        K_truncated = 0.5 * ScaHFX_ * JandK_truncated[1];
+
+      } else {
+        J_truncated = CalcERIs(TruncatedDensityMatrix, 1e-12);
+      }
+      double E_Hartree_truncated =
+          0.5 *
+          TruncatedDensityMatrix.cwiseProduct(J_truncated + K_truncated).sum();
+      // update the truncated Hamiltonian
+      Eigen::MatrixXd H_truncated =
+          H0_trunc_ + xc_truncated.matrix() + J_truncated + K_truncated;
+
+      double TruncatedTotalEnergy =
+          TruncatedDensityMatrix.cwiseProduct(H0_trunc_).sum() +
+          E_Hartree_truncated + E_xc_truncated - Initial_truncated_energy;
+      // ((TruncatedDensityMatrix - InitialActiveDmat_trunc_)
+      //      .cwiseProduct(v_embedding_trunc_)
+      //      .sum())
+      // get the new truncated density matrix
+      XTP_LOG(Log::info, *pLog_)
+          << std::endl
+          << "Electrons in = "
+          << TruncatedDensityMatrix.cwiseProduct(overlap.Matrix()).sum()
           << std::flush;
-      return false;
+      TruncatedDensityMatrix = conv_accelerator_.Iterate(
+          TruncatedDensityMatrix, H_truncated, MOs_trunc, TruncatedTotalEnergy);
+      XTP_LOG(Log::info, *pLog_)
+          << std::endl
+          << "Electrons out = "
+          << TruncatedDensityMatrix.cwiseProduct(overlap.Matrix()).sum()
+          << std::flush;
+
+      PrintMOs(MOs_trunc.eigenvalues(), Log::info);
+
+      if (conv_accelerator_.isConverged()) {
+        XTP_LOG(Log::error, *pLog_)
+            << TimeStamp() << " Total embedding energy has converged to "
+            << std::setprecision(9) << conv_accelerator_.getDeltaE()
+            << "[Ha] after " << this_iter + 1
+            << " iterations. DIIS error is converged up to "
+            << conv_accelerator_.getDIIsError() << std::flush;
+        XTP_LOG(Log::error, *pLog_)
+            << TimeStamp() << " Final Single Point Energy of embedding "
+            << std::setprecision(12) << TruncatedTotalEnergy << " Ha"
+            << std::flush;
+
+        PrintMOs(MOs_trunc.eigenvalues(), Log::error);
+        trunc_orb.setEmbeddedMOs(MOs_trunc);
+        trunc_orb.setNumofActiveElectrons(active_electrons_);
+        // CalcElDipole(trunc_orb);
```

Comment:
We need to fix this ElDipole call.

Reply:
Removed it for this branch. Will be fixed in the GWinMM embedding branch